### ISSUE TITLE
[temp.deduct.call] Is auto&& as a class template parameter supposed to be a forwarding reference

### ISSUE THREAD

**Issue by ranaanoop** [cplusplus/CWG#619]
Full name of submitter (unless configured in github; will be published with the issue): Anoop Rana

Reference (section label): [temp.deduct.call] 

Link to reflector thread (if any): https://stackoverflow.com/questions/79046493/perfect-forwarding-and-non-type-template-parameters

Issue description:

Consider the following example:
```
template<auto&& arg>
struct test {};
int x;    //global
test<x> t; //works
```
Is `arg` supposed to a universal/forwarding reference? While answering [this question](https://stackoverflow.com/questions/79046493/perfect-forwarding-and-non-type-template-parameters) I realized that the standard doesn't make this a universal reference but the cppreference and examples given there and  also the behavior of `auto&&` seems to make it a universal reference.

[temp.deduct.call] says:

> A forwarding reference is an rvalue reference to a cv-unqualified template parameter **that does not represent a template parameter of a class template** (during class template argument deduction ([over.match.class.deduct])). 

So clearly this is not considered to be a universal reference. 

But cppreference's [documentation](https://en.cppreference.com/w/cpp/language/reference#Forwarding_references) says the opposite:

> Forwarding references are a special kind of references that preserve the value category of a function argument, making it possible to forward it by means of std::forward. Forwarding references are either:
> * 1) function parameter of a function template declared as rvalue reference to cv-unqualified type template parameter of that same function template: 
>  * 2) **`auto&&` except when deduced from a brace-enclosed initializer list:**
> ```
> auto&& vec = foo();       // foo() may be lvalue or rvalue, vec is a forwarding reference
> ```

So, as you can see the cppreference's reference seems to say that it is a universal reference but the c++ standard doesn't. Also this change was added later into the standard. Before [temp.deduc.call] used to say "A forwarding reference is an rvalue reference to a cv-unqualified template parameter. "

So, is the class template parameter in the above example with `auto&&` supposed to be a universal/forwarding reference. Similarly, cppreference claims that `vec` is a forwarding reference. So is it too supposed to be a forwarding reference.

**Comment by jensmaurer** [comment]
This is not a forum to discuss what seems to be outdated cppreference documentation, a site entirely unrelated to the C++ standardization effort.

"So clearly this is not considered to be a universal reference."

So, where is the problem?

**Comment by ranaanoop** [comment]
@jensmaurer  You seem to miss the whole point. Why was this disallowed later on in the standard.

**Comment by jensmaurer** [comment]
"git blame" is your friend.  You can trace when and why the words were changed, and there should be a paper number or core issue attached to the change.

**Comment by t3nsor** [comment]
What the standard means to say is that if a function parameter has the form of a forwarding reference `T&&` but the function template is an implicit deduction guide and `T` was originally a template parameter of the *class template*, then it is not a forwarding reference. This wording was already in C++17, so it has not changed; there is no defect other than perhaps that the wording is hard to understand.

As for the given example, [temp.deduct.call] applies through [temp.arg.nontype]/1 and [dcl.type.auto.deduct]/3.

**Comment by ranaanoop** [comment]
@t3nsor  That is the point. As declared `template<auto&& arg> struct test {};`, we don't have a forwarding reference. But if we apply the deduction process like(invented function template and all), then since a function template is used instead of a class template, we will have a forwarding reference. But that is a hypothetical thing we did for knowing/deducing the type.  **It should not change the behavior/semantic/meaning of the program.** This still seems like a defect to me.

**Comment by jensmaurer** [comment]
What exactly is the example that would change meaning here?

**Comment by ranaanoop** [comment]
The change in meaning is "in concept". In particular, in the example given in the original issue, as declared `template<auto&& arg> struct test {};` we don't have a forwarding reference. But we know that when the argument is supplied like say `test<x> t;`(as in the original post/issue), `auto`, `auto&` etc works by the same process of template argument deduction. So we will have an invented/hypothetical "function template" like say `template<class T>
void f(T &&s);` which will be used to deduce the type of the passed argument. 

Basically, the hypothetical function template will involve a forwarding reference  which will make our original example of class template with `auto&&` behave like a forwarding reference even though it isn't(as declared).  

So there is mismatch in the behavior we're observing(which is forwarding) even though we don't actually have a forwarding reference. Basically the [duck test](https://en.wikipedia.org/wiki/Duck_test) : 
> "If it looks like a duck, swims like a duck, and quacks like a duck, then it probably is a duck."

**Comment by ranaanoop** [comment]
> this is not a standard about animals.

I never claimed it is. Ever heard of metaphors? If yes, then you wouldn't have said/made the first sentence in your reply. 

I would agree that this would be more appropriate for EWG etc(wherever more suitable) more than CWG(if that is what you also think) as what I want is to make the use of `auto&&` as given in our original issue, formally to be called a forwarding reference. So this is more like a proposal and hence not suitable here for CWG. So I will submit the issue at more suitable/appropriate place(like EWG etc). 

And again about my duck example/methphor, it was there to make a point. You could just have omitted the "standard is not about animal"  part as it was completely unnecessary imo.

**Comment by frederick-vs-ja** [comment]
> [temp.deduct.call] says:
> 
> > A forwarding reference is an rvalue reference to a cv-unqualified template parameter **that does not represent a template parameter of a class template** (during class template argument deduction ([over.match.class.deduct])).
> 
> So clearly this is not considered to be a universal reference.

@ranaanoop  No. You mistakenly ignored the "during class template argument deduction" part.

The restriction was added by [P0512R0](https://wg21.link/p0512r0) which made some fix for class template argument deduction, and `template<auto&&> ...` was added by [P0127R2](https://wg21.link/p0127r2) which wasn't directly related to CTAD.

So IIUC in this example `arg` is a forwarding reference.
```C++
template<auto&& arg>
struct test {};
int x;    //global
test<x> t; //works
```

> So, as you can see the cppreference's reference seems to say that it is a universal reference but the c++ standard doesn't.

No. The [page for CTAD](https://en.cppreference.com/w/cpp/language/class_template_argument_deduction) already says that the involved references don't accidently become forwarding reference.

cppreference is not outdated. It's just not consistent enough. I've [just fixed](https://en.cppreference.com/mwiki/index.php?title=cpp%2Flanguage%2Freference&diff=177058&oldid=171669) such inconsistency.
